modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/model/TrafficSource.java

```diff
@@ -39,13 +39,15 @@ public TrafficSource() {
 
 	public TrafficSource(
 		List<CountrySearchKeywords> countrySearchKeywordsList, String name,
-		long trafficAmount, double trafficShare) {
+		long trafficAmount, double trafficShare,
+		TrafficSourceDetails trafficSourceDetails) {
```

Comment:
Hey @darquesdev !!!

I have some concerns with enriching the model TrafficSource, as this represent the data we receive from AC. Would it better no keep two separate models?

One from our Frontend, and another one from the AC part?

The frontend could be called TrafficChannel, that is the current name we use

Reply:
As not all TrafficSource has CountrySearchKeywords or TrafficSourceDetails, I think we should think about using Types, only SearchTrafficSources should have countrySearchKeywordsList.

only ReferralTrafficSources (or another name) should have TrafficSourceDetails.

Reply:
Not sure where to put the boundary.. In `com.liferay.analytics.reports.web.internal.data.provider.AnalyticsReportsDataProvider#getTrafficSources` or in `com.liferay.analytics.reports.web.internal.portlet.action.GetTrafficSourcesMVCResourceCommand#_getTrafficSourcesJSONArray`?

Reply:
BTW, I don't see any TrafficChannel in the `code`

Reply:
If the transformation is made in `com.liferay.analytics.reports.web.internal.portlet.action.GetTrafficSourcesMVCResourceCommand#_getTrafficSourcesJSONArray` I could create a util class as happens with `com.liferay.segments.experiment.web.internal.util.SegmentsExperimentUtil` (see `com.liferay.segments.experiment.web.internal.util.SegmentsExperimentUtil#toSegmentsExperimentJSONObject`). In this class dependeing on the traffic source name the conversion to the specific `TrafficSource` (or `TrafficChannel`) JSONObject will be handled.

Reply:
`TrafficChannels` is the name we use in the UX, that is the reason I suggested to be used, but feel free to use any other name you think better suits the use case.

Use the simpler approach, what do you think is better?.

Now, we return all the data mixed in https://github.com/liferay/liferay-portal/blob/e15357135d1b8b08765c078b953d488f95051f9e/modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/data/provider/AnalyticsReportsDataProvider.java#L188.

Maybe we should replace this method for the new `getTrafficChannels`

And the method _getTraffiSources (that now it is private) should be moved as a public method. https://github.com/liferay/liferay-portal/blob/e15357135d1b8b08765c078b953d488f95051f9e/modules/dxp/apps/analytics/analytics-reports-web/src/main/java/com/liferay/analytics/reports/web/internal/data/provider/AnalyticsReportsDataProvider.java#L258

What do you think is the best approach?

Reply:
When you say `TrafficChannels` you mean `AcquisitionChannel`? And the approach you provided is to put the boundary in `AnalyticsReportsDataProvider`, right? I don't get why the _getTrafficSources method should become public

Reply:
No. I would create a new entity that represents the model we use to send data to the frontend, I would call it `TrafficChannel`, because it is not the AC name.

The`_getTrafficSources ` should be public because it is the direct call to AC, the same way we use with the other endpoints, like AdquisitionChannel.

Reply:
And the method getTrafficChannels should return a list of TrafficChannel. This new entity will be the base class which a specific class will inherit. Like ReferralTrafficChannel or SocialTrafficChannel. Right?

Reply:
I personally don't like base classes (inheritance would lead to undesired behaviors), I would recommend using an interface.

Reply:
But yeah, if this works for you I would recommend using types, for the definition of different behaviors.

Reply:
Ok, I'm trying to follow this approach ;) but I see a problem in the near future, as I'm designing every TrafficChannel will have a different `toJSONObject()` signature. For example the one for Search will require Locale but not in others. Maybe a refactor will simplify this aspect.

Reply:
I think that we could use also a Context object for this use case, also it can be a and interface, and every object could define their own context.

Reply:
I'm following the next:

- AnalyticsReportsDataProvider crete getTrafficChannels
- Create TrafficChannel Interface  (trafficAmont, trafficShare, name)
- Create classes for each type: SearchTrafficChannel e iclude detaild (countrySearchKeywords)
- toJSONObject(Context context) depending on the type of traffic channel
- Interzaf Context
- SearchTrafficChannelContext implmenent Context
- public Map<String, TrafficSource> getTrafficSources( -> getTrafficChannels
- private Map<String, TrafficSource> _getTrafficSources( -> PUBLIC
- Later see com.liferay.analytics.reports.web.internal.portlet.action.GetTrafficSourcesMVCResourceCommand#_getTrafficSourcesJSONArray -> _getTrafficSourcesJSONArray